File: shuvoice/cli/parser.py
"""CLI parser and compatibility mapping."""

from __future__ import annotations

import argparse
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from shuvoice.config import Config

LEGACY_FLAG_WARNING = "This legacy flag is deprecated and will be removed in a future release."

CONTROL_COMMAND_CHOICES = [
    "start",
    "stop",
    "toggle",
    "status",
    "ping",
    "metrics",
    "tts_speak",
    "tts_pause",
    "tts_resume",
    "tts_toggle_pause",
    "tts_restart",
    "tts_stop",
    "tts_status",
]


def _add_runtime_overrides(parser: argparse.ArgumentParser) -> None:
    parser.add_argument(
        "--asr-backend",
        choices=["nemo", "sherpa", "moonshine"],
        default=None,
        help="ASR backend selection (default: from config)",
    )
    parser.add_argument(
        "--device",
        default=None,
        help="Inference device (NeMo backend, default: from config)",
    )
    parser.add_argument(
        "--right-context",
        type=int,
        choices=[0, 1, 6, 13],
        default=None,
        help="NeMo streaming right context (0,1,6,13 => lower to higher latency/accuracy)",
    )
    parser.add_argument(
        "--sherpa-model-dir",
        default=None,
        help="Sherpa model directory containing tokens.txt + encoder/decoder/joiner ONNX files",
    )
    parser.add_argument(
        "--sherpa-model-name",
        default=None,
        help=(
            "Sherpa model archive name for auto-download when sherpa_model_dir is unset "
            "(default: from config)"
        ),
    )
    parser.add_argument(
        "--sherpa-provider",
        choices=["cpu", "cuda"],
        default=None,
        help="Sherpa execution provider (default: from config)",
    )
    parser.add_argument(
        "--sherpa-num-threads",
        type=int,
        default=None,
        help="Sherpa decoding threads (default: from config)",
    )
    parser.add_argument(
        "--sherpa-chunk-ms",
        type=int,
        default=None,
        help="Sherpa native chunk duration in milliseconds (default: from config)",
    )
    parser.add_argument(
        "--moonshine-model-name",
        default=None,
        help="Moonshine model name (for example: moonshine/tiny, moonshine/base)",
    )
    parser.add_argument(
        "--moonshine-model-dir",
        default=None,
        help="Local Moonshine model directory (encoder_model.onnx + decoder_model_merged.onnx)",
    )
    parser.add_argument(
        "--moonshine-model-precision",
        default=None,
        help="Moonshine model precision variant (default: from config)",
    )
    parser.add_argument(
        "--moonshine-chunk-ms",
        type=int,
        default=None,
        help="Moonshine native chunk duration in milliseconds (default: from config)",
    )
    parser.add_argument(
        "--moonshine-max-window-sec",
        type=float,
        default=None,
        help="Moonshine cumulative decode window in seconds (default: from config)",
    )
    parser.add_argument(
        "--moonshine-max-tokens",
        type=int,
        default=None,
        help="Moonshine max generated tokens per decode (default: from config)",
    )
    parser.add_argument(
        "--moonshine-provider",
        choices=["cpu", "cuda"],
        default=None,
        help="Moonshine execution provider (default: from config)",
    )
    parser.add_argument(
        "--moonshine-onnx-threads",
        type=int,
        default=None,
        help="Moonshine ONNX intra-op threads (0 = auto, default: from config)",
    )
    parser.add_argument(
        "--audio-device",
        default=None,
        help="Audio input device name or index (default: from config)",
    )
    parser.add_argument(
        "--input-gain",
        type=float,
        default=None,
        help="Multiply microphone PCM by this factor before ASR (default: from config)",
    )
    parser.add_argument(
        "--output-mode",
        choices=["final_only", "streaming_partial"],
        default=None,
        help="Text output mode (default: from config)",
    )
    parser.add_argument(
        "--control-socket",
        default=None,
        help="Override control socket path (default: $XDG_RUNTIME_DIR/shuvoice/control.sock)",
    )


def create_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(
        description="Streaming speech-to-text overlay for Hyprland",
    )
    parser.add_argument(
        "-v",
        "--verbose",
        action="store_true",
        help="Enable debug logging",
    )

    # Legacy top-level compatibility flags (one cycle).
    parser.add_argument(
        "--download-model",
        action="store_true",
        help=f"[legacy] Equivalent to `shuvoice model download`. {LEGACY_FLAG_WARNING}",
    )
    parser.add_argument(
        "--preflight",
        action="store_true",
        help=f"[legacy] Equivalent to `shuvoice preflight`. {LEGACY_FLAG_WARNING}",
    )
    parser.add_argument(
        "--list-audio-devices",
        action="store_true",
        help=(f"[legacy] Equivalent to `shuvoice audio list-devices`. {LEGACY_FLAG_WARNING}"),
    )
    parser.add_argument(
        "--wizard",
        action="store_true",
        help=f"[legacy] Equivalent to `shuvoice wizard`. {LEGACY_FLAG_WARNING}",
    )
    parser.add_argument(
        "--control",
        choices=CONTROL_COMMAND_CHOICES,
        default=None,
        help=f"[legacy] Equivalent to `shuvoice control <cmd>`. {LEGACY_FLAG_WARNING}",
    )
    parser.add_argument(
        "--control-wait-sec",
        type=float,
        default=2.0,
        help=(
            "When sending stop/toggle, wait up to this many seconds for post-stop "
            "processing to finish (0 disables wait)."
        ),
    )

    _add_runtime_overrides(parser)

    subparsers = parser.add_subparsers(dest="command")

    run_parser = subparsers.add_parser("run", help="Run the speech-to-text overlay")
    _add_runtime_overrides(run_parser)

    control_parser = subparsers.add_parser(
        "control", help="Send control command to running instance"
    )
    control_parser.add_argument(
        "control_command",
        choices=CONTROL_COMMAND_CHOICES,
        help="Control command",
    )
    control_parser.add_argument(
        "--control-wait-sec",
        type=float,
        default=2.0,
        help=(
            "When sending stop/toggle, wait up to this many seconds for post-stop "
            "processing to finish (0 disables wait)."
        ),
    )
    control_parser.add_argument(
        "--control-socket",
        default=None,
        help="Override control socket path (default: $XDG_RUNTIME_DIR/shuvoice/control.sock)",
    )

    preflight_parser = subparsers.add_parser("preflight", help="Run dependency and runtime checks")
    _add_runtime_overrides(preflight_parser)

    setup_parser = subparsers.add_parser(
        "setup",
        help="Bootstrap backend dependencies, model artifacts, and preflight checks",
    )
    _add_runtime_overrides(setup_parser)
    setup_parser.add_argument(
        "--install-missing",
        action="store_true",
        help="Attempt to install missing backend dependencies automatically",
    )
    setup_parser.add_argument(
        "--skip-model-download",
        action="store_true",
        help="Skip model download step during setup",
    )
    setup_parser.add_argument(
        "--skip-preflight",
        action="store_true",
        help="Skip final preflight checks",
    )
    setup_parser.add_argument(
        "--tts-local-voice",
        default=None,
        help=(
            "Curated Local Piper voice to download when tts_backend=local "
            "(for example: en_US-amy-medium)"
        ),
    )
    setup_parser.add_argument(
        "--tts-local-model-dir",
        default=None,
        help=(
            "Directory to store managed Local Piper voices when tts_backend=local "
            "(default: ~/.local/share/shuvoice/models/piper)"
        ),
    )
    setup_parser.add_argument(
        "--non-interactive",
        action="store_true",
        help="Disable interactive prompts and use recommended defaults where possible",
    )

    subparsers.add_parser("wizard", help="Launch the setup wizard")

    config_parser = subparsers.add_parser("config", help="Inspect and validate config")
    config_sub = config_parser.add_subparsers(dest="config_command")
    config_sub.add_parser("effective", help="Print merged effective config")
    config_sub.add_parser("path", help="Print active config file path")
    config_sub.add_parser("validate", help="Validate active config")
    config_set_parser = config_sub.add_parser("set", help="Set supported config keys")
    config_set_parser.add_argument(
        "key",
        choices=["typing_final_injection_mode"],
        help="Config key to set",
    )
    config_set_parser.add_argument(
        "value",
        choices=["auto", "clipboard", "direct"],
        help="New value for the selected key",
    )

    model_parser = subparsers.add_parser("model", help="Model management commands")
    _add_runtime_overrides(model_parser)
    model_sub = model_parser.add_subparsers(dest="model_command")
    model_sub.add_parser("download", help="Download model artifacts for active backend")

    audio_parser = subparsers.add_parser("audio", help="Audio utility commands")
    audio_sub = audio_parser.add_subparsers(dest="audio_command")
    audio_sub.add_parser("list-devices", help="List audio input devices")

    diagnostics_parser = subparsers.add_parser("diagnostics", help="Show runtime diagnostics")
    _add_runtime_overrides(diagnostics_parser)
    diagnostics_parser.add_argument(
        "--json",
        action="store_true",
        help="Output diagnostics as JSON",
    )

    return parser


def apply_cli_overrides(args: argparse.Namespace, config: Config) -> None:
    """Apply CLI argument overrides onto a Config instance."""
    if args.asr_backend:
        config.asr_backend = args.asr_backend
    if args.device:
        config.device = args.device
    if args.right_context is not None:
        config.right_context = int(args.right_context)
    if args.sherpa_model_dir is not None:
        config.sherpa_model_dir = args.sherpa_model_dir
    if args.sherpa_model_name is not None:
        config.sherpa_model_name = args.sherpa_model_name
    if args.sherpa_provider:
        config.sherpa_provider = args.sherpa_provider
    if args.sherpa_num_threads is not None:
        config.sherpa_num_threads = int(args.sherpa_num_threads)
    if args.sherpa_chunk_ms is not None:
        config.sherpa_chunk_ms = int(args.sherpa_chunk_ms)
    if args.moonshine_model_name is not None:
        config.moonshine_model_name = args.moonshine_model_name
    if args.moonshine_model_dir is not None:
        config.moonshine_model_dir = args.moonshine_model_dir
    if args.moonshine_model_precision is not None:
        config.moonshine_model_precision = args.moonshine_model_precision
    if args.moonshine_chunk_ms is not None:
        config.moonshine_chunk_ms = int(args.moonshine_chunk_ms)
    if args.moonshine_max_window_sec is not None:
        config.moonshine_max_window_sec = float(args.moonshine_max_window_sec)
    if args.moonshine_max_tokens is not None:
        config.moonshine_max_tokens = int(args.moonshine_max_tokens)
    if args.moonshine_provider:
        config.moonshine_provider = args.moonshine_provider
    if args.moonshine_onnx_threads is not None:
        config.moonshine_onnx_threads = int(args.moonshine_onnx_threads)
    if args.audio_device is not None:
        config.audio_device = (
            int(args.audio_device) if str(args.audio_device).isdigit() else args.audio_device
        )
    if args.input_gain is not None:
        config.input_gain = float(args.input_gain)
    if args.output_mode:
        config.output_mode = args.output_mode
    if args.control_socket:
        config.control_socket = args.control_socket


def resolve_command(
    args: argparse.Namespace,
    parser: argparse.ArgumentParser,
) -> tuple[str, list[str]]:
    """Resolve subcommand + legacy flag compatibility routing."""
    warnings: list[str] = []

    command = args.command
    if command == "control":
        args.control_action = args.control_command
        return "control", warnings
    if command == "preflight":
        return "preflight", warnings
    if command == "setup":
        return "setup", warnings
    if command == "wizard":
        return "wizard", warnings
    if command == "run":
        return "run", warnings
    if command == "config":
        if not args.config_command:
            parser.error("config subcommand required: effective | path | validate | set")
        return f"config_{args.config_command}", warnings
    if command == "model":
        if args.model_command != "download":
            parser.error("model subcommand required: download")
        return "model_download", warnings
    if command == "audio":
        if args.audio_command != "list-devices":
            parser.error("audio subcommand required: list-devices")
        return "audio_list_devices", warnings
    if command == "diagnostics":
        return "diagnostics", warnings

    # Legacy path.
    legacy_flags = [
        bool(args.preflight),
        bool(args.wizard),
        bool(args.download_model),
        bool(args.list_audio_devices),
        bool(args.control),
    ]
    if sum(int(v) for v in legacy_flags) > 1:
        parser.error("legacy flags are mutually exclusive")

    if args.control:
        args.control_action = args.control
        warnings.append("`--control` is deprecated; use `shuvoice control <cmd>`")
        return "control", warnings
    if args.preflight:
        warnings.append("`--preflight` is deprecated; use `shuvoice preflight`")
        return "preflight", warnings
    if args.wizard:
        warnings.append("`--wizard` is deprecated; use `shuvoice wizard`")
        return "wizard", warnings
    if args.download_model:
        warnings.append("`--download-model` is deprecated; use `shuvoice model download`")
        return "model_download", warnings
    if args.list_audio_devices:
        warnings.append("`--list-audio-devices` is deprecated; use `shuvoice audio list-devices`")
        return "audio_list_devices", warnings

    return "run", warnings

File: crates/shuvoice-cli/src/parser.rs
//! Clap CLI tree preserving ShuVoice's public command contract.

use std::path::PathBuf;

use clap::{Args, Parser, Subcommand, ValueEnum};

use crate::config::RuntimeOverrides;
use crate::control::ControlCmd;

pub const LEGACY_FLAG_WARNING: &str =
    "This legacy flag is deprecated and will be removed in a future release.";

#[derive(Debug, Clone, Parser)]
#[command(
    name = "shuvoice",
    about = "Streaming speech-to-text overlay for Hyprland",
    disable_version_flag = true
)]
pub struct Cli {
    /// Enable debug logging
    #[arg(short = 'v', long = "verbose", global = true)]
    pub verbose: bool,

    /// [legacy] Equivalent to `shuvoice model download`
    #[arg(long = "download-model", help = format!("[legacy] Equivalent to `shuvoice model download`. {LEGACY_FLAG_WARNING}"))]
    pub download_model: bool,

    /// [legacy] Equivalent to `shuvoice preflight`
    #[arg(long = "preflight", help = format!("[legacy] Equivalent to `shuvoice preflight`. {LEGACY_FLAG_WARNING}"))]
    pub legacy_preflight: bool,

    /// [legacy] Equivalent to `shuvoice audio list-devices`
    #[arg(long = "list-audio-devices", help = format!("[legacy] Equivalent to `shuvoice audio list-devices`. {LEGACY_FLAG_WARNING}"))]
    pub list_audio_devices: bool,

    /// [legacy] Equivalent to `shuvoice wizard`
    #[arg(long = "wizard", help = format!("[legacy] Equivalent to `shuvoice wizard`. {LEGACY_FLAG_WARNING}"))]
    pub legacy_wizard: bool,

    /// [legacy] Equivalent to `shuvoice control <cmd>`
    #[arg(long = "control", value_enum, help = format!("[legacy] Equivalent to `shuvoice control <cmd>`. {LEGACY_FLAG_WARNING}"))]
    pub legacy_control: Option<ControlCmd>,

    /// When sending stop/toggle, wait up to this many seconds for post-stop processing to finish (0 disables wait).
    #[arg(long = "control-wait-sec", default_value_t = 2.0)]
    pub control_wait_sec: f64,

    #[command(flatten)]
    pub overrides: RuntimeOverrideArgs,

    #[command(subcommand)]
    pub command: Option<Command>,
}

#[derive(Debug, Clone, Args, Default)]
pub struct RuntimeOverrideArgs {
    #[arg(long = "asr-backend", value_parser = ["nemo", "sherpa", "moonshine", "openai_realtime"])]
    pub asr_backend: Option<String>,
    #[arg(long = "device")]
    pub device: Option<String>,
    #[arg(long = "right-context", value_parser = parse_right_context)]
    pub right_context: Option<i64>,
    #[arg(long = "sherpa-model-dir")]
    pub sherpa_model_dir: Option<String>,
    #[arg(long = "sherpa-model-name")]
    pub sherpa_model_name: Option<String>,
    #[arg(long = "sherpa-provider", value_parser = ["cpu", "cuda"])]
    pub sherpa_provider: Option<String>,
    #[arg(long = "sherpa-num-threads")]
    pub sherpa_num_threads: Option<i64>,
    #[arg(long = "sherpa-chunk-ms")]
    pub sherpa_chunk_ms: Option<i64>,
    #[arg(long = "moonshine-model-name")]
    pub moonshine_model_name: Option<String>,
    #[arg(long = "moonshine-model-dir")]
    pub moonshine_model_dir: Option<String>,
    #[arg(long = "moonshine-model-precision")]
    pub moonshine_model_precision: Option<String>,
    #[arg(long = "moonshine-chunk-ms")]
    pub moonshine_chunk_ms: Option<i64>,
    #[arg(long = "moonshine-max-window-sec")]
    pub moonshine_max_window_sec: Option<f64>,
    #[arg(long = "moonshine-max-tokens")]
    pub moonshine_max_tokens: Option<i64>,
    #[arg(long = "moonshine-provider", value_parser = ["cpu", "cuda"])]
    pub moonshine_provider: Option<String>,
    #[arg(long = "moonshine-onnx-threads")]
    pub moonshine_onnx_threads: Option<i64>,
    #[arg(long = "audio-device")]
    pub audio_device: Option<String>,
    #[arg(long = "input-gain")]
    pub input_gain: Option<f64>,
    #[arg(long = "output-mode", value_parser = ["final_only", "streaming_partial"])]
    pub output_mode: Option<String>,
    #[arg(long = "control-socket")]
    pub control_socket: Option<String>,
}

fn parse_right_context(s: &str) -> Result<i64, String> {
    let v: i64 = s.parse().map_err(|e| format!("{e}"))?;
    if matches!(v, 0 | 1 | 6 | 13) {
        Ok(v)
    } else {
        Err("right-context must be one of: 0, 1, 6, 13".into())
    }
}

impl RuntimeOverrideArgs {
    pub fn to_overrides(&self) -> RuntimeOverrides {
        RuntimeOverrides {
            asr_backend: self.asr_backend.clone(),
            device: self.device.clone(),
            right_context: self.right_context,
            sherpa_model_dir: self.sherpa_model_dir.clone(),
            sherpa_model_name: self.sherpa_model_name.clone(),
            sherpa_provider: self.sherpa_provider.clone(),
            sherpa_num_threads: self.sherpa_num_threads,
            sherpa_chunk_ms: self.sherpa_chunk_ms,
            moonshine_model_name: self.moonshine_model_name.clone(),
            moonshine_model_dir: self.moonshine_model_dir.clone(),
            moonshine_model_precision: self.moonshine_model_precision.clone(),
            moonshine_chunk_ms: self.moonshine_chunk_ms,
            moonshine_max_window_sec: self.moonshine_max_window_sec,
            moonshine_max_tokens: self.moonshine_max_tokens,
            moonshine_provider: self.moonshine_provider.clone(),
            moonshine_onnx_threads: self.moonshine_onnx_threads,
            audio_device: self.audio_device.clone(),
            input_gain: self.input_gain,
            output_mode: self.output_mode.clone(),
            control_socket: self.control_socket.clone(),
        }
    }

    pub fn merge_into(&self, base: &mut RuntimeOverrides) {
        let o = self.to_overrides();
        if o.asr_backend.is_some() {
            base.asr_backend = o.asr_backend;
        }
        if o.device.is_some() {
            base.device = o.device;
        }
        if o.right_context.is_some() {
            base.right_context = o.right_context;
        }
        if o.sherpa_model_dir.is_some() {
            base.sherpa_model_dir = o.sherpa_model_dir;
        }
        if o.sherpa_model_name.is_some() {
            base.sherpa_model_name = o.sherpa_model_name;
        }
        if o.sherpa_provider.is_some() {
            base.sherpa_provider = o.sherpa_provider;
        }
        if o.sherpa_num_threads.is_some() {
            base.sherpa_num_threads = o.sherpa_num_threads;
        }
        if o.sherpa_chunk_ms.is_some() {
            base.sherpa_chunk_ms = o.sherpa_chunk_ms;
        }
        if o.moonshine_model_name.is_some() {
            base.moonshine_model_name = o.moonshine_model_name;
        }
        if o.moonshine_model_dir.is_some() {
            base.moonshine_model_dir = o.moonshine_model_dir;
        }
        if o.moonshine_model_precision.is_some() {
            base.moonshine_model_precision = o.moonshine_model_precision;
        }
        if o.moonshine_chunk_ms.is_some() {
            base.moonshine_chunk_ms = o.moonshine_chunk_ms;
        }
        if o.moonshine_max_window_sec.is_some() {
            base.moonshine_max_window_sec = o.moonshine_max_window_sec;
        }
        if o.moonshine_max_tokens.is_some() {
            base.moonshine_max_tokens = o.moonshine_max_tokens;
        }
        if o.moonshine_provider.is_some() {
            base.moonshine_provider = o.moonshine_provider;
        }
        if o.moonshine_onnx_threads.is_some() {
            base.moonshine_onnx_threads = o.moonshine_onnx_threads;
        }
        if o.audio_device.is_some() {
            base.audio_device = o.audio_device;
        }
        if o.input_gain.is_some() {
            base.input_gain = o.input_gain;
        }
        if o.output_mode.is_some() {
            base.output_mode = o.output_mode;
        }
        if o.control_socket.is_some() {
            base.control_socket = o.control_socket;
        }
    }
}

#[derive(Debug, Clone, Subcommand)]
pub enum Command {
    /// Run the speech-to-text overlay
    Run {
        #[command(flatten)]
        overrides: RuntimeOverrideArgs,
    },
    /// Send control command to running instance
    Control {
        #[arg(value_enum)]
        control_command: ControlCmd,
        #[arg(long = "control-wait-sec", default_value_t = 2.0)]
        control_wait_sec: f64,
        #[arg(long = "control-socket")]
        control_socket: Option<String>,
    },
    /// Run dependency and runtime checks
    Preflight {
        #[command(flatten)]
        overrides: RuntimeOverrideArgs,
    },
    /// Bootstrap backend dependencies, model artifacts, and preflight checks
    Setup {
        #[command(flatten)]
        overrides: RuntimeOverrideArgs,
        #[arg(long = "install-missing")]
        install_missing: bool,
        #[arg(long = "skip-model-download")]
        skip_model_download: bool,
        #[arg(long = "skip-preflight")]
        skip_preflight: bool,
        #[arg(long = "tts-local-voice")]
        tts_local_voice: Option<String>,
        #[arg(long = "tts-local-model-dir")]
        tts_local_model_dir: Option<PathBuf>,
        #[arg(long = "non-interactive")]
        non_interactive: bool,
    },
    /// Launch the setup wizard
    Wizard,
    /// Inspect and validate config
    Config {
        #[command(subcommand)]
        command: ConfigCommand,
    },
    /// Model management commands
    Model {
        #[command(flatten)]
        overrides: RuntimeOverrideArgs,
        #[command(subcommand)]
        command: ModelCommand,
    },
    /// Audio utility commands
    Audio {
        #[command(subcommand)]
        command: AudioCommand,
    },
    /// Show runtime diagnostics
    Diagnostics {
        #[command(flatten)]
        overrides: RuntimeOverrideArgs,
        #[arg(long = "json")]
        json: bool,
    },
}

#[derive(Debug, Clone, Subcommand)]
pub enum ConfigCommand {
    /// Print merged effective config
    Effective,
    /// Print active config file path
    Path,
    /// Validate active config
    Validate,
    /// Set supported config keys
    Set {
        #[arg(value_enum)]
        key: ConfigSetKey,
        #[arg(value_enum)]
        value: ConfigSetValue,
    },
}

#[derive(Debug, Clone, Copy, ValueEnum)]
pub enum ConfigSetKey {
    #[value(name = "typing_final_injection_mode")]
    TypingFinalInjectionMode,
    #[value(name = "typing_text_case")]
    TypingTextCase,
    #[value(name = "overlay_debug_mode")]
    OverlayDebugMode,
}

impl ConfigSetKey {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::TypingFinalInjectionMode => "typing_final_injection_mode",
            Self::TypingTextCase => "typing_text_case",
            Self::OverlayDebugMode => "overlay_debug_mode",
        }
    }
}

#[derive(Debug, Clone, Copy, ValueEnum)]
pub enum ConfigSetValue {
    Auto,
    Clipboard,
    Direct,
    Default,
    Lowercase,
    True,
    False,
}

impl ConfigSetValue {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Auto => "auto",
            Self::Clipboard => "clipboard",
            Self::Direct => "direct",
            Self::Default => "default",
            Self::Lowercase => "lowercase",
            Self::True => "true",
            Self::False => "false",
        }
    }
}

#[derive(Debug, Clone, Subcommand)]
pub enum ModelCommand {
    /// Download model artifacts for active backend
    Download,
}

#[derive(Debug, Clone, Subcommand)]
pub enum AudioCommand {
    /// List audio input devices
    #[command(name = "list-devices")]
    ListDevices,
}

/// Resolved internal route after legacy flag mapping.
#[derive(Debug, Clone)]
pub enum ResolvedCommand {
    Run {
        overrides: RuntimeOverrides,
    },
    Control {
        command: ControlCmd,
        wait_sec: f64,
        socket: Option<String>,
        overrides: RuntimeOverrides,
    },
    Preflight {
        overrides: RuntimeOverrides,
    },
    Setup {
        overrides: RuntimeOverrides,
        install_missing: bool,
        skip_model_download: bool,
        skip_preflight: bool,
        tts_local_voice: Option<String>,
        tts_local_model_dir: Option<PathBuf>,
        non_interactive: bool,
    },
    Wizard,
    ConfigPath,
    ConfigValidate,
    ConfigEffective,
    ConfigSet {
        key: ConfigSetKey,
        value: ConfigSetValue,
    },
    ModelDownload {
        overrides: RuntimeOverrides,
    },
    AudioListDevices,
    Diagnostics {
        overrides: RuntimeOverrides,
        json: bool,
    },
}

pub fn resolve_command(cli: &Cli) -> Result<(ResolvedCommand, Vec<String>), String> {
    let mut warnings = Vec::new();
    let mut root_overrides = cli.overrides.to_overrides();

    if let Some(command) = &cli.command {
        match command {
            Command::Run { overrides } => {
                overrides.merge_into(&mut root_overrides);
                return Ok((
                    ResolvedCommand::Run {
                        overrides: root_overrides,
                    },
                    warnings,
                ));
            }
            Command::Control {
                control_command,
                control_wait_sec,
                control_socket,
            } => {
                if let Some(sock) = control_socket {
                    root_overrides.control_socket = Some(sock.clone());
                }
                return Ok((
                    ResolvedCommand::Control {
                        command: *control_command,
                        wait_sec: *control_wait_sec,
                        socket: control_socket
                            .clone()
                            .or(root_overrides.control_socket.clone()),
                        overrides: root_overrides,
                    },
                    warnings,
                ));
            }
            Command::Preflight { overrides } => {
                overrides.merge_into(&mut root_overrides);
                return Ok((
                    ResolvedCommand::Preflight {
                        overrides: root_overrides,
                    },
                    warnings,
                ));
            }
            Command::Setup {
                overrides,
                install_missing,
                skip_model_download,
                skip_preflight,
                tts_local_voice,
                tts_local_model_dir,
                non_interactive,
            } => {
                overrides.merge_into(&mut root_overrides);
                return Ok((
                    ResolvedCommand::Setup {
                        overrides: root_overrides,
                        install_missing: *install_missing,
                        skip_model_download: *skip_model_download,
                        skip_preflight: *skip_preflight,
                        tts_local_voice: tts_local_voice.clone(),
                        tts_local_model_dir: tts_local_model_dir.clone(),
                        non_interactive: *non_interactive,
                    },
                    warnings,
                ));
            }
            Command::Wizard => return Ok((ResolvedCommand::Wizard, warnings)),
            Command::Config { command } => {
                let resolved = match command {
                    ConfigCommand::Effective => ResolvedCommand::ConfigEffective,
                    ConfigCommand::Path => ResolvedCommand::ConfigPath,
                    ConfigCommand::Validate => ResolvedCommand::ConfigValidate,
                    ConfigCommand::Set { key, value } => ResolvedCommand::ConfigSet {
                        key: *key,
                        value: *value,
                    },
                };
                return Ok((resolved, warnings));
            }
            Command::Model {
                overrides,
                command: ModelCommand::Download,
            } => {
                overrides.merge_into(&mut root_overrides);
                return Ok((
                    ResolvedCommand::ModelDownload {
                        overrides: root_overrides,
                    },
                    warnings,
                ));
            }
            Command::Audio {
                command: AudioCommand::ListDevices,
            } => return Ok((ResolvedCommand::AudioListDevices, warnings)),
            Command::Diagnostics { overrides, json } => {
                overrides.merge_into(&mut root_overrides);
                return Ok((
                    ResolvedCommand::Diagnostics {
                        overrides: root_overrides,
                        json: *json,
                    },
                    warnings,
                ));
            }
        }
    }

    // Legacy path.
    let legacy_flags = [
        cli.legacy_preflight,
        cli.legacy_wizard,
        cli.download_model,
        cli.list_audio_devices,
        cli.legacy_control.is_some(),
    ];
    if legacy_flags.iter().filter(|v| **v).count() > 1 {
        return Err("legacy flags are mutually exclusive".into());
    }

    if let Some(cmd) = cli.legacy_control {
        warnings.push("`--control` is deprecated; use `shuvoice control <cmd>`".into());
        return Ok((
            ResolvedCommand::Control {
                command: cmd,
                wait_sec: cli.control_wait_sec,
                socket: root_overrides.control_socket.clone(),
                overrides: root_overrides,
            },
            warnings,
        ));
    }
    if cli.legacy_preflight {
        warnings.push("`--preflight` is deprecated; use `shuvoice preflight`".into());
        return Ok((
            ResolvedCommand::Preflight {
                overrides: root_overrides,
            },
            warnings,
        ));
    }
    if cli.legacy_wizard {
        warnings.push("`--wizard` is deprecated; use `shuvoice wizard`".into());
        return Ok((ResolvedCommand::Wizard, warnings));
    }
    if cli.download_model {
        warnings.push("`--download-model` is deprecated; use `shuvoice model download`".into());
        return Ok((
            ResolvedCommand::ModelDownload {
                overrides: root_overrides,
            },
            warnings,
        ));
    }
    if cli.list_audio_devices {
        warnings
            .push("`--list-audio-devices` is deprecated; use `shuvoice audio list-devices`".into());
        return Ok((ResolvedCommand::AudioListDevices, warnings));
    }

    Ok((
        ResolvedCommand::Run {
            overrides: root_overrides,
        },
        warnings,
    ))
}
